File: app/lib/models/photo.dart
class Photo {
  final String id;
  final String filename;
  final String contentType;
  final String data; // Base64 encoded
  final String? description;
  final int fileSize;
  final DateTime dateCreation;

  Photo({
    required this.id,
    required this.filename,
    required this.contentType,
    required this.data,
    this.description,
    required this.fileSize,
    required this.dateCreation,
  });

  factory Photo.fromJson(Map<String, dynamic> json) {
    return Photo(
      id: json['_id'] ?? json['id'],
      filename: json['filename'],
      contentType: json['content_type'],
      data: json['data'],
      description: json['description'],
      fileSize: json['file_size'],
      dateCreation: DateTime.parse(json['date_creation']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'filename': filename,
      'content_type': contentType,
      'data': data,
      if (description != null) 'description': description,
    };
  }
}

class PhotoMetadata {
  final String id;
  final String filename;
  final String contentType;
  final String? description;
  final int fileSize;
  final DateTime dateCreation;

  PhotoMetadata({
    required this.id,
    required this.filename,
    required this.contentType,
    this.description,
    required this.fileSize,
    required this.dateCreation,
  });

  factory PhotoMetadata.fromJson(Map<String, dynamic> json) {
    return PhotoMetadata(
      id: json['_id'] ?? json['id'],
      filename: json['filename'],
      contentType: json['content_type'],
      description: json['description'],
      fileSize: json['file_size'],
      dateCreation: DateTime.parse(json['date_creation']),
    );
  }
}

File: backend/app/models/photo.py
"""Photo data models for storing product images."""
from datetime import datetime
from typing import Optional, Any, Annotated
from pydantic import BaseModel, Field, ConfigDict, BeforeValidator
from bson import ObjectId


def validate_object_id(v: Any) -> str:
    """Validate and convert ObjectId to string."""
    if isinstance(v, ObjectId):
        return str(v)
    if isinstance(v, str):
        if not ObjectId.is_valid(v):
            raise ValueError("Invalid ObjectId")
        return v
    raise ValueError("Invalid ObjectId type")


# Create an annotated type for MongoDB ObjectId
PyObjectId = Annotated[str, BeforeValidator(validate_object_id)]


class PhotoBase(BaseModel):
    """Base photo model for uploading."""
    filename: str = Field(..., description="Original filename")
    content_type: str = Field(..., description="MIME type (e.g., image/jpeg)")
    data: str = Field(..., description="Base64 encoded image data")
    description: Optional[str] = Field(default=None, description="Photo description")


class PhotoCreate(PhotoBase):
    """Schema for creating a new photo."""
    pass


class PhotoResponse(BaseModel):
    """Response model for photo data."""
    id: PyObjectId = Field(alias="_id", description="Photo ID")
    filename: str = Field(..., description="Original filename")
    content_type: str = Field(..., description="MIME type")
    data: str = Field(..., description="Base64 encoded image data")
    description: Optional[str] = Field(default=None, description="Photo description")
    file_size: int = Field(..., description="File size in bytes")
    date_creation: datetime = Field(..., description="Upload timestamp")
    
    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        json_encoders={ObjectId: str}
    )


class PhotoMetadata(BaseModel):
    """Photo metadata without the actual image data (for listings)."""
    id: PyObjectId = Field(alias="_id", description="Photo ID")
    filename: str = Field(..., description="Original filename")
    content_type: str = Field(..., description="MIME type")
    description: Optional[str] = Field(default=None, description="Photo description")
    file_size: int = Field(..., description="File size in bytes")
    date_creation: datetime = Field(..., description="Upload timestamp")
    
    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        json_encoders={ObjectId: str}
    )
